homeassistant/components/unifi/entity.py

```diff
@@ -0,0 +1,176 @@
+            )
+        )
+
+        # Config entry options updated
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass,
+                self.controller.signal_options_update,
+                self.async_signal_options_updated,
+            )
+        )
+
+        # Subscribe to events if defined
+        if description.event_to_subscribe is not None:
+            self.async_on_remove(
+                self.controller.api.events.subscribe(
+                    self.async_event_callback,
+                    description.event_to_subscribe,
+                )
+            )
+
+        self.async_custom_subscribe()
+
+    @callback
+    def async_signalling_callback(self, event: ItemEvent, obj_id: str) -> None:
+        """Update the entity state."""
+        if event == ItemEvent.DELETED and obj_id == self._obj_id:
+            self.hass.async_create_task(self.remove_item({self._obj_id}))
+            return
+
+        description = self.entity_description
+        if not description.supported_fn(self.controller, self._obj_id):
+            self.hass.async_create_task(self.remove_item({self._obj_id}))
+            return
+
+        self._attr_available = description.available_fn(self.controller, self._obj_id)
+        self.async_update_state(event, obj_id)
+        self.async_write_ha_state()
+
+    @callback
+    def async_signal_reachable_callback(self) -> None:
+        """Call when controller connection state change."""
+        self.async_signalling_callback(ItemEvent.ADDED, self._obj_id)
+
+    async def async_signal_options_updated(self) -> None:
+        """Config entry options are updated, remove entity if option is disabled."""
+        if not self.entity_description.allowed_fn(self.controller, self._obj_id):
+            await self.remove_item({self._obj_id})
+
+    async def remove_item(self, keys: set) -> None:
+        """Remove entity if object ID is part of set."""
+        if self._obj_id not in keys or self._removed:
+            return
+        self._removed = True
+        if self.registry_entry:
+            er.async_get(self.hass).async_remove(self.entity_id)
+        else:
+            await self.async_remove(force_remove=True)
+
+    @callback
+    def async_initiate_state(self) -> None:
+        """Initiate entity state.
+
+        Perform additional actions setting up platform entity child class state.
+        """
+
+    @callback
+    def async_update_state(self, event: ItemEvent, obj_id: str) -> None:
+        """Update entity state.
+
+        Perform additional actions updating platform entity child class state.
+        """
+
+    @callback
+    def async_event_callback(self, event: Event) -> None:
+        """Update entity state based on subscribed event.
+
+        Perform additional action updating platform entity child class state.
+        """
+
+    @callback
+    def async_custom_subscribe(self) -> None:
+        """Do custom subscriptions."""
```

Comment:
Should these be abstractmethods, or is it expected that some derived entities don't implement all of these?

Reply:
`async_initiate_state` and `async_update_state` are needed to be implemented in all platforms whereas `async_event_callback` and `async_custom_subscribe` are optional, how would you suggest it be done?
Do you have an example on how to use them?

Reply:
If `async_initiate_state` and `async_update_state` are needed to be implemented, you should let `UnifiEntity` inherit from ABC and mark `async_initiate_state` and `async_update_state` as abstractmethods. That will prevent instances of classes derived from `UnifiEntity` to be instantiated if those methods are not implemented. Entity already inherits ABC, so there's no need to add it again to `UnifiEntity`, so it would be something like this:

```py
from abc import abstractmethod

class UnifiEntity(Entity, Generic[HandlerT, DataT]):

    @callback
    @abstractmethod
    def async_initiate_state(self) -> None:
        ...

    @callback
    @abstractmethod
    def async_update_state(self, event: ItemEvent, obj_id: str) -> None:
        ...
```

Reply:
Thanks, done.